### PR TITLE
define EReal log and exp

### PR DESCRIPTION
**Pull request by RemyDegenne**

### CODE REVIEW

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+/-- Logarithm as a function from `ℝ≥0∞` to `EReal`. -/
+noncomputable
+def log : ℝ≥0∞ → EReal
+| ∞ => ⊤
+| x => if x = 0 then ⊥ else Real.log x.toReal
+
+@[simp] lemma log_zero : log 0 = ⊥ := by simp [log]
+@[simp] lemma log_top : log ∞ = ⊤ := by simp [log]
+
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean:39]:
```suggestion
@[simp] lemma log_zero : log 0 = ⊥ := if_pos rfl
@[simp] lemma log_top : log ∞ = ⊤ := rfl
```

Comment (RemyDegenne) [reply TestingLowerBounds/ForMathlib/ERealLogExp.lean:39]: I'll change it! That suggestion is not at the right line, so I cannot apply it through the github interface though.

Comment (LorenzoLuccioli) [reply TestingLowerBounds/ForMathlib/ERealLogExp.lean:39]: Ok thanks, I'm trying to do the review directly from VScode, so I can run the code, but I still have to figure out exactly how the comments behave.

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+  exact not_le.mpr hx
+
+@[simp] lemma log_eq_bot_iff {x : ℝ≥0∞} : log x = ⊥ ↔ x = 0 := by
+  refine ⟨fun h ↦ ?_, fun hx ↦ by simp [hx]⟩
+  by_contra hx_ne_zero
+  by_cases hx_top : x = ∞
+  · simp [hx_top] at h
+  have hx_pos : 0 < x.toReal := ENNReal.toReal_pos hx_ne_zero hx_top
+  rw [← ENNReal.ofReal_toReal hx_top, log_ofReal_of_pos hx_pos] at h
+  simp at h
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
  exact coe_ne_bot _ h
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  have hx_pos : 0 < x.toReal := ENNReal.toReal_pos hx_zero hx_ne_top
+  rw [← ENNReal.ofReal_toReal hx_ne_top, log_ofReal_of_pos hx_pos] at h
+  simp at h
+
+lemma log_mono {a b : ℝ≥0∞} (h : a ≤ b) : log a ≤ log b := by
+  by_cases hb_top : b = ∞
+  · simp [hb_top]
+  have ha_ne_top : a ≠ ∞ := by
+    refine fun ha ↦ hb_top (top_le_iff.mp ?_)
+    rwa [ha] at h
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
  have ha_ne_top : a ≠ ∞ := fun ha ↦ hb_top (top_le_iff.mp <| ha ▸ h)
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  by_cases ha_zero : a = 0
+  · simp [ha_zero]
+  have hb_ne_zero : b ≠ 0 := by
+    refine fun hb ↦ ha_zero (le_antisymm ?_ zero_le')
+    rwa [hb] at h
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
  have hb_ne_zero : b ≠ 0 := fun hb ↦ ha_zero (le_antisymm (hb ▸ h) zero_le')
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+lemma log_inv (x : ℝ≥0∞) : log x⁻¹ = - log x := by
+  by_cases hx_top : x = ∞
+  · simp [hx_top]
+  by_cases hx_zero : x = 0
+  · simp [hx_zero]
+  have hx_pos : 0 < x.toReal := ENNReal.toReal_pos hx_zero hx_top
+  rw [← ENNReal.ofReal_toReal hx_top, log_ofReal_of_pos hx_pos, ← ENNReal.ofReal_inv_of_pos hx_pos,
+    log_ofReal_of_pos (by positivity)]
+  norm_cast
+  rw [Real.log_inv]
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
  rw [← ENNReal.ofReal_toReal hx_top, log_ofReal_of_pos hx_pos, ← ENNReal.ofReal_inv_of_pos hx_pos,
    log_ofReal_of_pos (by positivity), Real.log_inv]
  norm_cast
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  · simp [hb_zero]
+  by_cases ha_top : a = ∞
+  · rw [ha_top, log_top, EReal.top_add_of_ne_bot, ENNReal.top_mul hb_zero, log_top]
+    rwa [ne_eq, log_eq_bot_iff]
+  by_cases hb_top : b = ∞
+  · rw [hb_top, log_top, EReal.add_top_of_ne_bot, ENNReal.mul_top ha_zero, log_top]
+    rwa [ne_eq, log_eq_bot_iff]
+  have ha_pos : 0 < a.toReal := ENNReal.toReal_pos ha_zero ha_top
+  have hb_pos : 0 < b.toReal := ENNReal.toReal_pos hb_zero hb_top
+  rw [← ENNReal.ofReal_toReal ha_top,← ENNReal.ofReal_toReal hb_top,
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
  rw [← ENNReal.ofReal_toReal ha_top, ← ENNReal.ofReal_toReal hb_top,
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+/-- Exponential as a function from `EReal` to `ℝ≥0∞`. -/
+noncomputable
+def exp : EReal → ℝ≥0∞
+| ⊥ => 0
+| ⊤ => ∞
+| (x : ℝ) => ENNReal.ofReal (Real.exp x)
+
+@[simp] lemma exp_bot : exp ⊥ = 0 := by simp [exp]
+@[simp] lemma exp_top : exp ⊤ = ∞ := by simp [exp]
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
@[simp] lemma exp_bot : exp ⊥ = 0 := rfl
@[simp] lemma exp_top : exp ⊤ = ∞ := rfl
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+@[simp] lemma exp_coe (x : ℝ) : exp x = ENNReal.ofReal (Real.exp x) := by
+  have h : (x : EReal) = some (some x) := rfl
+  simp [h, exp]
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
@[simp] lemma exp_coe (x : ℝ) : exp x = ENNReal.ofReal (Real.exp x) := rfl
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+    · simp
+  · rw [top_le_iff] at h
+    simp [h.symm, exp_top, top_le_iff]
+
+lemma exp_neg (x : EReal) : exp (-x) = (exp x)⁻¹ := by
+  induction' x using EReal.rec with x
+  · simp
+  · rw [exp_coe, ← EReal.coe_neg, exp_coe, ← ENNReal.ofReal_inv_of_pos (Real.exp_pos _)]
+    congr
+    rw [Real.exp_neg]
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
  · rw [exp_coe, ← EReal.coe_neg, exp_coe, ← ENNReal.ofReal_inv_of_pos (Real.exp_pos _), 
      Real.exp_neg]
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+@[simp] lemma exp_log (x : ℝ≥0∞) : exp (log x) = x := by
+  by_cases hx_top : x = ∞
+  · simp [hx_top]
+  by_cases hx_zero : x = 0
+  · simp [hx_zero]
+  have hx_pos : 0 < x.toReal := ENNReal.toReal_pos hx_zero hx_top
+  rw [← ENNReal.ofReal_toReal hx_top, log_ofReal_of_pos hx_pos, exp_coe, Real.exp_log hx_pos]
+
+@[simp]
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean:188]: Is there a reason for the `@[simp]` to be on the line above the statement some times, and on the same line of the statement other times?

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+lemma continuous_log : Continuous log := logOrderIso.continuous
+
+lemma continuous_exp : Continuous exp := logOrderIso.symm.continuous
+
+lemma measurable_log : Measurable log := continuous_log.measurable
+
+lemma measurable_exp : Measurable exp := continuous_exp.measurable
+
+lemma _root_.Measurable.ereal_log {α : Type*} {_ : MeasurableSpace α}
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean:215]: Is there a reason for `MeasurableSpace` not to be stated as an instance?
```suggestion
lemma _root_.Measurable.ereal_log {α : Type*} [MeasurableSpace α]
```
Same for the lemma below (`Measurable.ereal_exp`)

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+  · rfl
+  · rw [ENNReal.toReal_ofReal]
+    exact (not_le.mp h_nonpos).le
+
+lemma log_ofReal_of_pos {x : ℝ} (hx : 0 < x) : log (ENNReal.ofReal x) = Real.log x := by
+  rw [log_ofReal, if_neg]
+  exact not_le.mpr hx
+
+@[simp] lemma log_eq_bot_iff {x : ℝ≥0∞} : log x = ⊥ ↔ x = 0 := by
+  refine ⟨fun h ↦ ?_, fun hx ↦ by simp [hx]⟩
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
  refine ⟨fun h ↦ ?_, fun hx ↦ hx ▸ log_zero⟩
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+lemma _root_.Measurable.ereal_log {α : Type*} {_ : MeasurableSpace α}
+    {f : α → ℝ≥0∞} (hf : Measurable f) :
+    Measurable fun x ↦ log (f x) := measurable_log.comp hf
+
+lemma _root_.Measurable.ereal_exp {α : Type*} {_ : MeasurableSpace α}
+    {f : α → EReal} (hf : Measurable f) :
+    Measurable fun x ↦ exp (f x) := measurable_exp.comp hf
+
+end LogExp
+
```
Comment (LorenzoLuccioli) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean:224]: With `logOrderIso` we basically get the metrizability of `EReal` for free, should we add it here?

```suggestion

instance : TopologicalSpace.MetrizableSpace EReal :=
  logOrderIso.symm.toHomeomorph.embedding.metrizableSpace

```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean (hunk 11 above)
Comment (RemyDegenne) [reply TestingLowerBounds/ForMathlib/ERealLogExp.lean:215]: Yes there is a reason! The measurable space has to be the one of the measurability hypothesis `hf`. With a typeclass argument, the lemma becomes applicable only to the measurable space inferred by the typeclass mechanism. With an implicit argument, I can pass `hf` of the shape `measurable[m] f` for some sub-sigma-algebra `m` and the lemma applies.

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+section Log
+
+/-- Logarithm as a function from `ℝ≥0∞` to `EReal`. -/
+noncomputable
+def log : ℝ≥0∞ → EReal
+| ∞ => ⊤
+| x => if x = 0 then ⊥ else Real.log x.toReal
+
+@[simp] lemma log_zero : log 0 = ⊥ := by simp [log]
+@[simp] lemma log_top : log ∞ = ⊤ := by simp [log]
```
Comment (RemyDegenne) [inline TestingLowerBounds/ForMathlib/ERealLogExp.lean]:
```suggestion
@[simp] lemma log_zero : log 0 = ⊥ := if_pos rfl
@[simp] lemma log_top : log ∞ = ⊤ := rfl
```

File: TestingLowerBounds/ForMathlib/ERealLogExp.lean (hunk 10 above)
Comment (RemyDegenne) [reply TestingLowerBounds/ForMathlib/ERealLogExp.lean:188]: I put it on the same line sometimes to gain space. When the whole lemma is one line, it's not great to double the space it takes with one line for simp. It's not very important.

**Review by LorenzoLuccioli — APPROVED** [review]

**Pull request merged**